### ISSUE TITLE
System.Net.Http.HttpRequestException: The SSL connection could not be established

### ISSUE THREAD

**Issue by RobinMennens-UiPath** [snowflakedb/snowflake-connector-net#685]
1. What version of .NET driver are you using?
2.0.22
   
2. What operating system and processor architecture are you using?
Any CPU. Trying to change underlying base docker image.
   
3. What version of .NET framework are you using?
  .NET 7
   C# 11  
   
4. What did you do?

We're trying to update the base image of our docker file. We used to be on a base image that has `OpenSSL 1.1.1`. Now we're moving to `RHEL9` (Red Hat Enterprise Linux 9) which does not support OpenSSL 1.1.1 anymore but has `OpenSSL 3`. Our hunch is that this is causing the issue described below.
After updating, we're not able to connect to Snowflake anymore. The connector continuously keeps trying to connect until it reaches a timeout. We're constantly seeing a `System.Net.Http.HttpRequestException: The SSL connection could not be established` exception  and then a `Response returned was null.`

Full exception:
```
System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception.
---> System.Security.Authentication.AuthenticationException: The remote certificate is invalid because of errors in the certificate chain: RevocationStatusUnknown, OfflineRevocation
at System.Net.Security.SslStream.SendAuthResetSignal(ProtocolToken message, ExceptionDispatchInfo exception)
at System.Net.Security.SslStream.CompleteHandshake(SslAuthenticationOptions sslAuthenticationOptions)
at System.Net.Security.SslStream.ForceAuthenticationAsync[TIOAdapter](Boolean receiveFirst, Byte[] reAuthenticationData, CancellationToken cancellationToken)
at System.Net.Http.ConnectHelper.EstablishSslConnectionAsync(SslClientAuthenticationOptions sslOptions, HttpRequestMessage request, Boolean async, Stream stream, CancellationToken cancellationToken)
--- End of inner exception stack trace ---
at System.Net.Http.ConnectHelper.EstablishSslConnectionAsync(SslClientAuthenticationOptions sslOptions, HttpRequestMessage request, Boolean async, Stream stream, CancellationToken cancellationToken)
at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
at System.Net.Http.HttpConnectionPool.AddHttp11ConnectionAsync(QueueItem queueItem)
at System.Threading.Tasks.TaskCompletionSourceWithCancellation`1.WaitWithCancellationAsync(CancellationToken cancellationToken)
at System.Net.Http.HttpConnectionPool.HttpConnectionWaiter`1.WaitForConnectionAsync(Boolean async, CancellationToken requestCancellationToken)
at System.Net.Http.HttpConnectionPool.SendWithVersionDetectionAndRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
at System.Net.Http.DiagnosticsHandler.SendAsyncCore(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
at System.Net.Http.DecompressionHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
at Snowflake.Data.Core.HttpUtil.RetryHandler.SendAsync(HttpRequestMessage requestMessage, CancellationToken cancellationToken)
```

5. What did you expect to see?

 Connecting to Snowflake worked before but now it does not anymore.

**Comment by sfc-gh-dszmolka** [comment]
hi and thank you for submitting this issue ! so I'm trying to reproduce without success so far, here's what I'm doing:
1. run `registry.access.redhat.com/ubi9` 
```
# grep ^PRETTY /etc/*release
/etc/os-release:PRETTY_NAME="Red Hat Enterprise Linux 9.2 (Plow)"
```
RHEL9 indeed.
2. confirm this has OpenSSL v3
```
# openssl version
OpenSSL 3.0.7 1 Nov 2022 (Library: OpenSSL 3.0.7 1 Nov 2022)
```
3. install `dotnet-sdk-7.0`
```
# dotnet --list-runtimes
Microsoft.AspNetCore.App 7.0.7 [/usr/lib64/dotnet/shared/Microsoft.AspNetCore.App]
Microsoft.NETCore.App 7.0.7 [/usr/lib64/dotnet/shared/Microsoft.NETCore.App]
```
4. installed `mono` so I could run `nuget.exe`, which was particularly fun 😅  and in hindsight, probably totally unnecessary because as it turned out, the Snowflake .NET Connector can be conveniently installed from `dotnet` CLI as well:
```
dotnet add package Snowflake.Data --version 2.0.25
```
5. initialized the project with `dotnet new console`, then using the following sophisticated program for testing:
```c#
using System;
using System.Data;
using System.Data.Common;
using Snowflake.Data.Client;
namespace SnowflakeTestProgram 
{
    class Program
    {
        private static readonly log4net.ILog log = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);
        static void Main(string[] args)
        {
            try
            {
                using (IDbConnection conn = new SnowflakeDbConnection())
                {
                    conn.ConnectionString = "account=myaccount.west-europe.azure; user=admin; password=mypassword; DB=TEST_DB; SCHEMA=PUBLIC";
                    conn.Open();
                    Console.WriteLine("Connection successful!");
                    using (IDbCommand cmd = conn.CreateCommand())
                    {
                        cmd.CommandText = "select current_user()"; 
                        IDataReader reader = cmd.ExecuteReader();
                        while (reader.Read())
                        {
                            Console.WriteLine(reader.GetString(0));
                        }
                        conn.Close();
                    }
                }
            }
            catch (DbException exc)
            {
                Console.WriteLine("Error Message: {0}", exc.Message);
            }
        }
    }
}

```

Getting the following result:
```
# dotnet run
Connection successful!
ADMIN
```

so apparently it can connect to Snowflake and run the simple query.

Now, I can imagine your environment and even the program you're using might be wildly different, so as of next step, do you think it would be possible to
* share the exact build steps so I could attempt to set up an environment as close to yours as possible
* share the runnable program which when executed, leads to the issue you're seeing
* perhaps using additional TLS debug steps like the ones described in https://tatham.blog/2007/05/23/debugging-ssltls-problems-in-net/ and https://techcommunity.microsoft.com/t5/iis-support-blog/use-system-net-trace-and-ssl-alert-protocol-to-troubleshoot-ssl/ba-p/287709 (or something similar) can provide further clues

**Comment by sfc-gh-dszmolka** [comment]
all right, this is great news, thank you for letting me know ! since the problem is fixed with the latest (2.0.25) version of the driver, i'm closing this one out.
